### PR TITLE
feat(eval): Add `HotpotQA` benchmark in `distractor` setting

### PR DESCRIPTION
**Pull request by jon-chuang**
# Description

Notebook: https://github.com/jerryjliu/llama_index/blob/31c329620f85f47ac997b1ac7a6d978bb56571ee/docs/examples/evaluation/HotpotQADistractor.ipynb

## Type of Change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [X] Added new notebook (that tests end-to-end)

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: llama_index/evaluation/benchmarks/hotpotqa.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+class HotpotQAEvaluator:
+    """
+    Refer to https://hotpotqa.github.io/ for more details on the dataset
+    """
+
+    def __init__(self) -> None:
+        try:
+            pass
+        except ImportError:
+            raise ImportError(
```
Comment (jon-chuang) [inline llama_index/evaluation/benchmarks/hotpotqa.py:30]: get rid

File: llama_index/query_engine/retriever_query_engine.py [hunk 2]
```diff
@@ -105,24 +105,34 @@ def from_args(
         )
 
     def _apply_node_postprocessors(
-        self, nodes: List[NodeWithScore]
+        self, nodes: List[NodeWithScore], query_bundle: QueryBundle
     ) -> List[NodeWithScore]:
         for node_postprocessor in self._node_postprocessors:
-            nodes = node_postprocessor.postprocess_nodes(nodes)
+            nodes = node_postprocessor.postprocess_nodes(
+                nodes, query_bundle=query_bundle
```
Comment (logan-markewich) [inline llama_index/query_engine/retriever_query_engine.py:112]: good catch! Forgot to add this

File: llama_index/evaluation/benchmarks/hotpotqa.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+import tqdm
+import requests
+import json
+from collections import Counter
+
+DEV_DISTRACTOR_URL = """http://curtis.ml.cmu.edu/datasets/\
+hotpot/hotpot_dev_distractor_v1.json"""
+
+
+class HotpotQAEvaluator:
```
Comment (logan-markewich) [inline llama_index/evaluation/benchmarks/hotpotqa.py:21]: I wonder if the BierEvaluator and the HotpotQAEvaluator should be sharing some common interface/base class?

File: docs/examples/evaluation/HotpotQADistractor.ipynb
Comment (logan-markewich) [inline docs/examples/evaluation/HotpotQADistractor.ipynb:1]: Some of the answers here are kind of funny. Even the first sample question about nationality, it's like... what? 😆

File: llama_index/evaluation/benchmarks/hotpotqa.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            query_engine, RetrieverQueryEngine
+        ), "query_engine must be a RetrieverQueryEngine for this evaluation"
+        retriever = HotpotQARetriever(query_objects)
+        # Mock the query engine with a retriever
+        query_engine = query_engine.with_retriever(retriever=retriever)
+
+        scores = {"exact_match": 0.0, "f1": 0.0}
+
+        for query in query_objects:
+            response = query_engine.query(query["question"])
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply llama_index/evaluation/benchmarks/hotpotqa.py:96]: Oh I see, I didn't realize this was a possibility. Will investigate further in the notebook!

File: docs/examples/evaluation/HotpotQADistractor.ipynb
Comment (jon-chuang) [reply docs/examples/evaluation/HotpotQADistractor.ipynb:1]: Yes, exactly, I was reflecting that it's far from "clean" data for sure.. q5 makes 0 sense in its phrasing for instance.

Comment (jon-chuang) [reply docs/examples/evaluation/HotpotQADistractor.ipynb:1]: It's sad that such a popular dataset is not actually super clean.

File: llama_index/evaluation/benchmarks/hotpotqa.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jon-chuang) [reply llama_index/evaluation/benchmarks/hotpotqa.py:96]: Oh right, I mean, in this particular case, the retriever is mocked, so it requires an exact match on the question. 

I will try to use `custom_embedding_strs` using  a `QueryBundle`, but it's a bit of a hack. But maybe it makes sense for this particuar benchmark.

[the remark this answers is not in the document]
Comment (jon-chuang) [reply llama_index/evaluation/benchmarks/hotpotqa.py]: Ok, the performance improved with the hack. Kinda happy about that.

On the flip side, the explainability went way down. 

Would really like to see that separation between answer and explanation.

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**